File: zephyr_one/mobile/android/feature-connections/src/main/kotlin/one/zephyr/mobile/feature/connections/TcpReachabilityTester.kt
package one.zephyr.mobile.feature.connections

import java.net.InetSocketAddress
import java.net.Socket
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import one.zephyr.mobile.model.Connection
import one.zephyr.mobile.model.MobileError

/**
 * Local TCP reachability. Independent of any protocol engine and of the main end.
 *
 * Measures the time to complete a TCP handshake to [Connection.host]:[Connection.port]. That is
 * the "测试连接延迟" the editor already knows how to render as [ConnectionTestResult.Reachable].
 * Authentication is a different question and stays with the protocol engines.
 */
class TcpReachabilityTester(
    private val timeoutMs: Int = DEFAULT_TIMEOUT_MS,
    private val nowMs: () -> Long = System::currentTimeMillis,
    private val connect: (String, Int, Int) -> Unit = { host, port, timeout ->
        Socket().use { socket ->
            socket.connect(InetSocketAddress(host, port), timeout)
        }
    },
) : ConnectionTester {

    override suspend fun test(
        connection: Connection,
        credentials: ConnectionTestCredentials,
    ): ConnectionTestResult = withContext(Dispatchers.IO) {
        val host = connection.host.trim()
        if (host.isEmpty()) {
            return@withContext ConnectionTestResult.Failed(
                MobileError.local("test_no_host", "没有填写主机", retryable = false),
            )
        }
        if (connection.port !in 1..65_535) {
            return@withContext ConnectionTestResult.Failed(
                MobileError.local("test_bad_port", "端口无效", retryable = false),
            )
        }
        val started = nowMs()
        try {
            connect(host, connection.port, timeoutMs)
            ConnectionTestResult.Reachable((nowMs() - started).coerceAtLeast(0L))
        } catch (error: Throwable) {
            ConnectionTestResult.Failed(
                MobileError.local(
                    code = "test_unreachable",
                    message = error.message?.takeIf { it.isNotBlank() } ?: "无法连接到 ${host}:${connection.port}",
                    retryable = true,
                ),
            )
        }
    }

    companion object {
        const val DEFAULT_TIMEOUT_MS = 5_000
    }
}

File: zephyr_one/mobile/ios/Sources/ZephyrUI/TcpReachabilityTester.swift
import Foundation
#if canImport(Network)
import Network
#endif

/// Local TCP reachability. Independent of any protocol engine and of the
/// main end. Measures the time to complete a TCP handshake.
public struct TcpReachabilityTester: ConnectionTester, @unchecked Sendable {
    public var timeoutMs: Int
    public var nowMs: @Sendable () -> Int64
    public var connect: @Sendable (String, Int, Int) async throws -> Void

    public init(
        timeoutMs: Int = 5_000,
        nowMs: @escaping @Sendable () -> Int64 = {
            Int64(Date().timeIntervalSince1970 * 1_000)
        },
        connect: @escaping @Sendable (String, Int, Int) async throws -> Void = { host, port, timeoutMs in
            try await TcpReachabilityTester.open(host: host, port: port, timeoutMs: timeoutMs)
        }
    ) {
        self.timeoutMs = timeoutMs
        self.nowMs = nowMs
        self.connect = connect
    }

    public func test(_ connection: Connection) async throws -> ConnectionTestResult {
        let host = connection.host.trimmingCharacters(in: .whitespacesAndNewlines)
        if host.isEmpty {
            return .failed(MobileError.local(code: "test_no_host", message: "没有填写主机"))
        }
        if connection.port < 1 || connection.port > 65_535 {
            return .failed(MobileError.local(code: "test_bad_port", message: "端口无效"))
        }
        let started = nowMs()
        do {
            try await connect(host, connection.port, timeoutMs)
            return .reachable(roundTripMs: max(0, nowMs() - started))
        } catch {
            return .failed(
                MobileError.local(
                    code: "test_unreachable",
                    message: error.localizedDescription,
                    retryable: true
                )
            )
        }
    }

    public static func open(host: String, port: Int, timeoutMs: Int) async throws {
        #if canImport(Network)
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            let connection = NWConnection(
                host: NWEndpoint.Host(host),
                port: NWEndpoint.Port(rawValue: UInt16(port)) ?? .any,
                using: .tcp
            )
            let finisher = TcpConnectionFinisher(connection: connection, continuation: continuation)
            connection.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    finisher.finish(.success(()))
                case let .failed(error):
                    finisher.finish(.failure(error))
                case .cancelled:
                    finisher.finish(.failure(CancellationError()))
                default:
                    break
                }
            }
            connection.start(queue: DispatchQueue.global(qos: .userInitiated))
            DispatchQueue.global(qos: .userInitiated).asyncAfter(
                deadline: .now() + .milliseconds(timeoutMs)
            ) {
                finisher.finish(.failure(NSError(domain: NSPOSIXErrorDomain, code: Int(ETIMEDOUT))))
            }
        }
        #else
        throw NSError(domain: NSPOSIXErrorDomain, code: Int(ENOTSUP))
        #endif
    }
}

#if canImport(Network)
private final class TcpConnectionFinisher: @unchecked Sendable {
    private let lock = NSLock()
    private let connection: NWConnection
    private let continuation: CheckedContinuation<Void, Error>
    private var finished = false

    init(connection: NWConnection, continuation: CheckedContinuation<Void, Error>) {
        self.connection = connection
        self.continuation = continuation
    }

    func finish(_ result: Result<Void, Error>) {
        lock.lock()
        guard !finished else {
            lock.unlock()
            return
        }
        finished = true
        lock.unlock()
        connection.cancel()
        continuation.resume(with: result)
    }
}
#endif
